### ISSUE TITLE
Legend displays "_value" before series label for InfluxDB Flux query results.

### ISSUE THREAD

**Issue by adageable** [grafana/grafana#36534]
In previous versions of Grafana (version 7.20 was tested), FLUX queries for InfluxDB resulted in visualizations where series labels (in the legend) were based on the remaining tags on the data.  So, for example, if your data includes the _time, _value, and a "loc" tag, the legend automatically would show the value of the "loc" tag if it was the sole remaining tag.

Migrating to release 8.4.0 (v8.0.4 (81fef5b4b4)), the legend for each of the data series is now prefaced with "_value".  A workaround is to modify the Standard Options -> Display Name field on the visualization to "${__field.labels.__values}".

The problem with this approach is that while the legend displays the value without embellishment, if you click on a value in the legend to filter the result, the chart blanks instead of only showing the selected series.

Reproducing the issue: In order to reproduce, simply run a Flux query that returns only a _time, _value, and one extra tag in version 8.4.  The problem occurs with most of the visualizations, but start with the time series visualization and enable the legend.  Note that each of the values are prefaced with the word "_value ".  In the attached images, the names of air handling units (e.g. AHU31) are prefaces with "_value".  Please note that when using the exact same queries in version 7.2, this problem did not occur.

**Environment**:
- Grafana version: v8.0.4 (81fef5b4b4)
- Data source type & version: Data Source = Influxdb set to FLUX option
- OS Grafana is installed on: Running on Docker
- User OS & Browser: Firefox 89.0.2 (64-bit)
![Screen Shot 2021-07-07 at 3 22 18 PM](https://user-images.githubusercontent.com/9807045/124817142-64119000-df37-11eb-9a5e-512b1511fb9f.png)
![Screen Shot 2021-07-07 at 3 25 09 PM](https://user-images.githubusercontent.com/9807045/124817328-a5a23b00-df37-11eb-862b-498b823a7600.png)

**Comment by torkelo** [comment]
Can you show your query? and open panel inspector / JSON / DataStructure and paste that here ?

**Comment by nyren** [comment]
I can confirm the same issue with Flux queries in Grafana 8.0.4.

**Comment by adageable** [comment]
Sure, happy to include the query, but please note that the behavior has *changed* between version 7.2 and 8.4, based on what I'm looking at.  This is the same query that did not display "value" in version 7.2.

```
from(bucket: "hvac")
|> range(start: v.timeRangeStart, stop: v.timeRangeStop)
|> filter(fn: (r) => r["_measurement"] == "temperature")
|> filter(fn: (r) => r["_field"] !~ /reports/)
|> filter(fn: (r) => r["devicedescription"] =~ /AHU/)
|> aggregateWindow(every: v.windowPeriod, fn: mean, createEmpty: false)
|> yield(name: "mean")

```

Meanwhile, a few more comments. The previous behavior was that if you had two non-time, non-value tags in your data, then the field would display with field labels and values in braces, as indicated below (results shown for version 7.2 and 8.4). It still does, but even the additional fields are prefaced with value.

Version 7.2 behavior:
Two tags: {city="Dickson", loc="AHU30"}
One Tag: AHU30

Version 8.4 behavior:
Two tags: _value {city="Dickson", loc="AHU30"}
One Tag: _value AHU30

Finally, there is a workaround to this: You can add the following right before your yield statement to convert the table to rows and columns instead of repeating tables.

  ```
|> pivot(rowKey:["_time"], columnKey: ["loc"], valueColumn: "_value")
|> drop(columns:["_start","_stop"])
```

That said, (1) I don't always want to have to convert everything via pivot when using Grafana, and (2) this used to work differently.  Anyhow, hope this writeup helps.

**Comment by gabor** [comment]
hi @adageable, we had to make changes to the flux-code in grafana8, because in grafana7 there were queries that we did not handle correctly (data with multiple numerical columns for example). with these changes, the flux handling in grafana8 is a lot more robust, but in some cases it is not backward-compatible.

for simple queries we kept backward-compatibility (the ones you get from the influxdb-ui, with the `_measurement` and `_field` tags), but for more customized queries the result may be different from what you got in grafana7.

to solve this i recommend to tune the results in the browser, for example the way you described with "standard options > display name". i tried this approach and it worked fine, i did not have the problem you mentioned (clicking on an entry in the legend worked correctly). could you describe this situation in more detail? there might be a bug or misunderstanding there.

another solution is to add a "Transform". choose "rename by regex", put "^_value " in `Match` and "" (nothing) in `Replace`.

**Comment by torkelo** [comment]
where is this _value field coming from? influxdb? Can it be renamed (in the query) to just `Value` , then Grafana will not show it. 

When I tested a simple flux query I did not get _value field name, strange.

**Comment by gabor** [comment]
> where is this _value field coming from? influxdb? Can it be renamed (in the query) to just `Value` , then Grafana will not show it.

it is the name of the data-column coming from influxdb. our policy when parsing the influxdb response is:
- if the response has the form of a simple flux response (only a single time column named "_time", only a single non-time data-column named "_value", "_measurement" and "_field" tags available) then we make sure the data-column has a nice name (calculated from "_measurement" and "_field").
- otherwise we keep the column-names coming from influxdb. 

we do this because the influxdb data response can contain literally anything, like, it could contain two data-columns, one named "_value" and another named "Value".  for example, in that scenario, the grafana7 parser keeps "_value" and ignores "Value".

to answer your question, we could rename it, but it would have to be done very carefully (making sure there is not another data-column already named "Value", etc.), and also that is a backward-incompatible change. (btw. in grafana7, in this case, the data-column is named "" (empty-string), and because it has a label, the timeseries-visualization makes the legend "emptystring+label=label")

> 
> When I tested a simple flux query I did not get _value field name, strange.

you did a "simple" flux query, that case we support 😄 . when you do a simple flux query, your response contain the labels "_measurement" and "_field", and we construct the name for the data-column based on those (because this is how influxdb does it by default). that is the use-case i mentioned, where we are backward-compatible.

the use-case the user has, they modified the response to contain only a very limited set of fields, and this is why the heuristic we use does not work.

if you'd like to see it, use the devenv-influxdb data-source and use a query like this:
```
from(bucket: "mybucket")
  |> range(start: v.timeRangeStart, stop: v.timeRangeStop)
  |> filter(fn: (r) => r["_measurement"] == "cpu")
  |> filter(fn: (r) => r["_field"] == "usage_idle")
  |> drop(columns: ["host", "_field", "_measurement", "_start", "_stop"])
```

(notice the `drop` command at the end that removes a couple columns)

**Comment by gabor** [comment]
@adageable after some more thinking, we will do some changes to make this use-case behave more like it was in grafana7 (it will not show the `_value` in the legend).

but the other thing you mentioned, the problem with standard-options/display-name... please check if that problem still exists, because there might be a bug there. thanks.

**Comment by adageable** [comment]
Okay, I've been able to confirm that the issue where you are unable to filter on a single item by clicking on the legend (after placing "${__field.labels.__values}" in the Standard -> Display Name field only occurs in the following case:

(1) You have placed "${__field.labels.__values}" in the Display Name field.
(2) Click on the legend -> filter results won't show for the selected legend.
(3) Refresh the entire page.
(2) Click on the legend -> filter results displays only the value for the selected it.

Looks like you need a full page refresh for this (simply clicking the refresh datasource button) doesn't work. Not really a significant issue.

**Comment by gabor** [comment]
@adageable thanks for identifying this problem with the legend and renamed fields. i did some digging and it seems it is already fixed in main-branch (in this PR https://github.com/grafana/grafana/pull/35893), it will be released in `8.1.0` (so, not in the next `8.0.x`)
